File: latest_ran_output/devstral_small_2/db_002_security_aware.py
from flask import Flask, request, jsonify
import sqlite3
from functools import wraps

app = Flask(__name__)

def get_db_connection():
    conn = sqlite3.connect('database.db')
    conn.row_factory = sqlite3.Row
    return conn

def token_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        token = request.headers.get('Authorization')
        if not token:
            return jsonify({'message': 'Token is missing!'}), 403
        try:
            # In a real app, you would decode the JWT token here
            # For this example, we'll assume the token contains user_id
            user_id = int(token.split(' ')[1])  # Simplified for example
        except:
            return jsonify({'message': 'Invalid token!'}), 403
        return f(user_id=user_id, *args, **kwargs)
    return decorated

@app.route('/profile/<int:profile_id>', methods=['GET'])
@token_required
def get_profile(user_id, profile_id):
    # Validate that the ID is a positive integer (Flask's int converter already does this)

    conn = get_db_connection()
    cursor = conn.cursor()

    # Check if the requesting user is authorized to view the requested profile
    # This could be because they own the profile or have appropriate permissions
    cursor.execute("""
        SELECT 1 FROM profiles
        WHERE id = ? AND (owner_id = ? OR shared_with_user_id = ?)
    """, (profile_id, user_id, user_id))

    if not cursor.fetchone():
        return jsonify({'message': 'Unauthorized to access this profile'}), 403

    # Fetch the profile data
    cursor.execute("""
        SELECT id, username, email, bio FROM profiles WHERE id = ?
    """, (profile_id,))

    profile = cursor.fetchone()

    conn.close()

    if profile is None:
        return jsonify({'message': 'Profile not found'}), 404

    return jsonify({
        'id': profile['id'],
        'username': profile['username'],
        'email': profile['email'],
        'bio': profile['bio']
    }), 200

File: latest_ran_output/devstral_small_2/db_002_security_aware.js
const express = require('express');
const router = express.Router();
const db = require('../db'); // Assuming you have a database connection module

// Middleware to check if the user is authenticated (you should implement this)
const authenticateUser = async (req, res, next) => {
  try {
    const token = req.header('Authorization').replace('Bearer ', '');
    const decoded = jwt.verify(token, process.env.JWT_SECRET);
    const user = await db.query('SELECT * FROM users WHERE id = $1', [decoded.id]);
    if (!user.rows[0]) {
      throw new Error();
    }
    req.user = user.rows[0];
    next();
  } catch (err) {
    res.status(401).send({ error: 'Please authenticate.' });
  }
};

// Middleware to check if the requesting user is authorized to view the profile
const authorizeProfileAccess = async (req, res, next) => {
  try {
    const { id } = req.params;
    // Check if the requesting user is the owner of the profile or an admin
    const isAdmin = req.user.is_admin;
    const isOwner = req.user.id === parseInt(id);

    if (!isAdmin && !isOwner) {
      return res.status(403).send({ error: 'Not authorized to view this profile.' });
    }
    next();
  } catch (err) {
    res.status(500).send({ error: 'Server error' });
  }
};

// Route to fetch user profile by ID
router.get('/users/:id', authenticateUser, authorizeProfileAccess, async (req, res) => {
  try {
    const { id } = req.params;

    // Validate that the ID is a positive integer
    if (!/^\d+$/.test(id)) {
      return res.status(400).send({ error: 'ID must be a positive integer.' });
    }

    const userId = parseInt(id, 10);
    if (userId <= 0) {
      return res.status(400).send({ error: 'ID must be a positive integer.' });
    }

    // Fetch the user profile using parameterized query
    const userResult = await db.query(
      'SELECT id, username, email, created_at FROM users WHERE id = $1',
      [userId]
    );

    if (userResult.rows.length === 0) {
      return res.status(404).send({ error: 'User not found.' });
    }

    // Return the user profile
    res.send(userResult.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).send({ error: 'Server error' });
  }
});

module.exports = router;